### PR TITLE
Add support for inline_asm_elementwise in Inductor lowerings

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #129950
* __->__ #129846

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

This doesn't actually expose `inline_asm_elementwise` from any public API, but makes it pretty easy to register a lowering for a custom op that uses it.

image

### CODE REVIEW

File: test/inductor/test_custom_lowering.py
Comment (yanboliang) [inline test/inductor/test_custom_lowering.py:139]: You want to register ```add_custom_lowering````?

File: torch/_inductor/codegen/triton.py
Comment (shunting314) [inline torch/_inductor/codegen/triton.py:729]: Do we want to expose pack as an parameter to `inline_asm_elementwise`?

File: torch/_inductor/codegen/triton.py
Comment (eellison) [inline torch/_inductor/codegen/triton.py:725]: What about all the other arguments:

```
Parameters
:
asm – assembly to run. Must match target’s assembly format.

constraints – asm constraints in [LLVM format](https://llvm.org/docs/LangRef.html#inline-asm-constraint-string)

args – the input tensors, whose values are passed to the asm block

dtype – the element type(s) of the returned tensor(s)

is_pure – if true, the compiler assumes the asm block has no side-effects

pack – the number of elements to be processed by one instance of inline assembly

_builder – the builder
```

Would it make sense to expose any of them in api ?

**Review by shunting314 — APPROVED** [review]
What are the general uses cases of using inline assembly like this?

Is this to cover the cases where triton does not have a corresponding op in its TTIR?

File: torch/_inductor/kernel/flex_attention.py
Comment (drisspg) [inline torch/_inductor/kernel/flex_attention.py:550]: dont land this .. haha

File: torch/_inductor/codegen/triton.py
Comment (Chillee) [reply torch/_inductor/codegen/triton.py:725]: To be clear, this is currently not a "public" API - you can only access it by registering a custom lowering and calling `ops`. In principle, some of these might make sense, but I don't really understand what most of them would be used for, so I only expose `asm` and `dtype` for now.

**Comment by Chillee** [comment]
@shunting314 Yeah, there a variety of `ptx` instructions that may not be explicitly exposed to the user. For example, in this case, `libdevice.tanh` gets lowered to this in SASS

image

And this causes a 2.5x slowdown in FlashAttention.

OTOH, if you use `tanh.approx`, there's only about a 15% slowdown.

File: torch/_inductor/codegen/triton.py
Comment (Chillee) [reply torch/_inductor/codegen/triton.py:725]: Maybe I should just expose them in-case somebody wants to use them 🤔

**Comment by shunting314** [comment]
> OTOH, if you use tanh.approx, there's only about a 15% slowdown.

BTW, tanh.approx returns an approximation? Does that hurt accuracy or it's negligible?

**Comment by Chillee** [comment]
@shunting314 Yeah, it definitely hurts accuracy. But in cases like this, users might want to trade off a 2.5x perf improvement for degraded accuracy.

**Comment by jeffdaily** [comment]
Reverted because ROCm does not support CUDA-specific inline asm.  If it were a single unit test that was added by this PR, we would have opened an issue to disable the test.  We try to avoid opening multiple DISABLED issues when they can be root-caused back to a single PR like this.  This PR introduces 2 unit tests.  I can understand you might not be able to modify these tests to also work correctly for ROCm, so in this case please add `@skipCUDAIfRocm` decorator to these tests.  I have also added the ciflow/rocm label to run the full ROCm CI which caught this after the merge.

**Comment by Chillee** [comment]
Oh oops, should have thought about that.

**Pull request closed without merging**
